cmd/thanos/query.go

```diff
@@ -91,6 +91,9 @@ func registerQuery(m map[string]setupFunc, app *kingpin.Application, name string
 	enablePartialResponse := cmd.Flag("query.partial-response", "Enable partial response for queries if no partial_response param is specified.").
 		Default("true").Bool()
 
+	storeReceiveTimeout := modelDuration(cmd.Flag("store.receive-timeout", "Maximum time to wait for any data from store. If Store doesn't send any data any storeReceiveTimeout the Store will be igored and partial data will be returned.").
```

Comment:
so I would try to be as close as possible to Golang HTTP timeouts. In my opinion, the `storeReadTimeout` we have here  https://github.com/improbable-eng/thanos/pull/920 is actually `storeGRPCRequest` timeout wheras this is actually `storeReadTimeout` as `net/http` Golang lib's [ReadTimeout](https://golang.org/pkg/net/http/#Server),, right?

Reply:
```suggestion
	storeReceiveTimeout := modelDuration(cmd.Flag("store.receive-timeout", "Maximum time to wait for any data from store. If Store doesn't send any data any storeReceiveTimeout the Store will be ignored and partial data will be returned.").
```

Reply:
Hey @povilasv 
I would offer to keep one `store.read-timeout` from here https://github.com/improbable-eng/thanos/pull/895 that could be a bit less than `query.timeout`
(just to have a time gap for query evaluation).
Also it would be cool to change approach on receiving data from stores: 
I would use common channel for receiving data from all stores and merge series only when reading finished.

Reply:
hey @d-ulyanov I think the same as  @bwplotka your  `read-timeout` is acually grpc call timeout.
TBH I don't see how it differs from `--query.timeout`, as end user I would expect for query to timeout in duration defined in `--query.timeout`, so maybe they should be married into one?

Re my timeout is actually `read-timeout`, similiar to HTTP. 

Re changing approach to a single channel I wouldn't agree on that, because we would have no way to timeout slow Store API's, but let's talk about it more, maybe I'm not getting something.

Reply:
I agree, for end user it would be very comfortable to deal with 2 parameters `query.timeout` and `partial-response`.
But I don't see any way how we can make it work at the moment, because `query.timeout` is closing inside prometheus engine and we can't handle it properly. 
I think in future `query.timeout` and `store.read-timeout` should be married.

In our infrastructure we already have 11 stores and need this feature to provide stable single endpoint for metrics for developers - from time to time some stores could be failed by different reasons. And we're not expecting that one slow store can fail whole request :(

About single channel - I've meant that we can 1) all stores can independently write response to one common channel (we already have such buffer in code) 2) we read from this channel to buffer in goroutine.

In this case we still can have independent store.read-timeout for all stores.
Now mergedSet inits with `a.Next()`, `b.Next()` and it leads to block if store `a` works slowly :(

Reply:
My current read time out is completely different form query timeout. Read time is about Store API not responding, query timeout is about killing query if it takes too long. So no way they could be married. Just look at HTTP timeouts.

Although your timeout is about killing query if it takes too long so as --query.timeout

Reply:
Also my timeout should be specified in milliseconds, like 200 - 300ms is a good value.
Query timeout could be 10mins. How are you going to merge that?